shared/dataflow/codeql/dataflow/VariableCapture.qll

```diff
@@ -0,0 +1,930 @@
+  }
+
+  /**
+   * Holds if `access` is a reference to `ce` evaluated in the `i`th node of `bb`.
+   * The reference is restricted to be in the same callable as `ce` as a
+   * precaution, even though this is expected to hold for all the given aliased
+   * accesses.
+   */
+  private predicate localClosureAccess(ClosureExpr ce, Expr access, BasicBlock bb, int i) {
+    ce.hasAliasedAccess(access) and
+    access.hasCfgNode(bb, i) and
+    pragma[only_bind_out](bb.getEnclosingCallable()) =
+      pragma[only_bind_out](closureExprGetCallable(ce))
+  }
+
+  /**
+   * Holds if we need an additional read of `v` in the `i`th node of `bb` in
+   * order to synchronize the value stored on `closure`.
+   * `topScope` is true if the read is in the defining callable of `v`.
+   *
+   * Side-effects of potentially calling `closure` at this point will be
+   * observed in a similarly synthesized post-update node for this read of `v`.
+   */
+  private predicate synthRead(
+    CapturedVariable v, BasicBlock bb, int i, boolean topScope, Expr closure
+  ) {
+    exists(ClosureExpr ce | closureCaptures(ce, v) |
+      ce.hasCfgNode(bb, i) and ce = closure
+      or
+      localClosureAccess(ce, closure, bb, i)
+    ) and
+    if v.getCallable() != bb.getEnclosingCallable() then topScope = false else topScope = true
+  }
+
+  /**
+   * Holds if there is an access of a captured variable inside a closure in the
+   * `i`th node of `bb`, such that we need to synthesize a `this.` qualifier.
+   */
+  private predicate synthThisQualifier(BasicBlock bb, int i) {
+    synthRead(_, bb, i, false, _) or
+    captureRead(_, bb, i, false, _) or
+    captureWrite(_, bb, i, false, _)
+  }
+
+  private newtype TCaptureContainer =
+    TVariable(CapturedVariable v) or
+    TThis(Callable c) { captureAccess(_, c) }
+
+  /**
+   * A storage location for a captured variable in a specific callable. This is
+   * either the variable itself (in its defining scope) or an instance variable
+   * `this` (in a capturing scope).
+   */
+  private class CaptureContainer extends TCaptureContainer {
+    string toString() {
+      exists(CapturedVariable v | this = TVariable(v) and result = v.toString())
+      or
+      result = "this" and this = TThis(_)
+    }
+  }
+
+  /** Holds if `cc` needs a definition at the entry of its callable scope. */
+  private predicate entryDef(CaptureContainer cc, BasicBlock bb, int i) {
+    exists(Callable c |
+      entryBlock(bb) and
+      pragma[only_bind_out](bb.getEnclosingCallable()) = c and
+      i =
+        -1 +
+          min(int j |
+            j = 1 or
+            captureRead(_, bb, j, _, _) or
+            captureWrite(_, bb, j, _, _) or
+            synthRead(_, bb, j, _, _)
+          )
+    |
+      cc = TThis(c)
+      or
+      exists(CapturedParameter p | cc = TVariable(p) and p.getCallable() = c)
+    )
+  }
+
+  private module CaptureSsaInput implements Ssa::InputSig {
+    class BasicBlock instanceof Input::BasicBlock {
```

Comment:
Does `final class BasicBlock = Input::BasicBlock;` not work?

Reply:
It does! That's much better. This has been work-in-progress for a while, and I think that feature wasn't ready when I first wrote this.